Project/Source/input.c

```diff
@@ -0,0 +1,310 @@
+  L3GD20_FilterStruct.HighPassFilter_CutOff_Frequency = L3GD20_HPFCF_0;						// 7.2 Hz (max)
+  L3GD20_FilterConfig(&L3GD20_FilterStruct) ;
+  
+  L3GD20_FilterCmd(L3GD20_HIGHPASSFILTER_ENABLE);
+}
+
+void EXTI2_IRQHandler(void)
+{
+	if (EXTI_GetITStatus(EXTI_Line2) != RESET)
+	{
+		EXTI_ClearITPendingBit(EXTI_Line2);
+		queue(EXTI2_IRQn);
+		
+		//test
+		gpio_led_toggle(1);
+	}
+}
+
+void EXTI3_IRQHandler(void)
+{
+	if (EXTI_GetITStatus(EXTI_Line3) != RESET)
+	{
+		EXTI_ClearITPendingBit(EXTI_Line3);
+		queue(EXTI3_IRQn);
+		
+		//test
+		gpio_led_toggle(2);
+	}
+}
+
+
+void EXTI4_IRQHandler(void)
+{
+	if (EXTI_GetITStatus(EXTI_Line4) != RESET)
+	{
+		EXTI_ClearITPendingBit(EXTI_Line4);
+		queue(EXTI4_IRQn);
+		
+		//za test bez timera - kad treca tipka oslobada prve dvije:
+		dequeue();
+		//promijeni trenutni blok - inace radi timer
+		wait_queue ^= 0x02;
+		
+	}
+}
+
+static void queue(uint32_t irq){
+	uint8_t top = 0;
+	
+	NVIC_DisableIRQ(irq);
+	
+	switch (irq){
+		case EXTI2_IRQn:
+			top = top | 0x80;
+			break;
+		case EXTI3_IRQn:
+			top |= 0x40;
+			break;
+		case EXTI4_IRQn:
+			top |= 0x20;
+			break;
+	}
+	
+	//puni sljedeci blok
+	if (!(wait_queue & 0x02))
+		top = top >> 3;
+	
+	wait_queue |= top;
+	
+	//pokreni timer na zadani delay
+}
+
+static void dequeue(void){
+	uint8_t top = wait_queue;
+	
+	if (!(wait_queue & 0x02)){
+			top = top << 3;
+			wait_queue &= 0xE3;
+	}
+	else wait_queue &= 0x1F;
+	
+	if (top & 0x80){
+		EXTI_ClearITPendingBit(EXTI_Line2);
+		NVIC_EnableIRQ(EXTI2_IRQn);
+	}
+	if (top & 0x40){
+		EXTI_ClearITPendingBit(EXTI_Line3);
+		NVIC_EnableIRQ(EXTI3_IRQn);
+	}
+	if (top & 0x20){
+		EXTI_ClearITPendingBit(EXTI_Line4);
+		NVIC_EnableIRQ(EXTI4_IRQn);
+	}
+	
+}
+
+void gpio_led_toggle(uint8_t LED_ID)
+{
+	switch(LED_ID)
+	{
+		case LED3_GREEN_ID:
+			GPIO_ToggleBits(LED3_GREEN_GPIOx, LED3_GREEN_PinNumber);
+			break;
+		case LED4_RED_ID:
+			GPIO_ToggleBits(LED4_RED_GPIOx, LED4_RED_PinNumber);
+			break;
+	}
+}
+
+
+void read_ang_rate (float* pfData)
+{
+  uint8_t tmpbuffer[4] ={0};
+  int16_t tmpdata[2] = {0};
+  int i =0;
+  
+	L3GD20_Read(tmpbuffer,L3GD20_OUT_X_L_ADDR,4);			
+	// L3GD20_OUT_X_L_ADDR = 28h ; read 28h - 2Bh
+  
+	for(i=0; i<2; i++)
+    {
+      tmpdata[i]=(int16_t)(((uint16_t)tmpbuffer[2*i+1] << 8) + tmpbuffer[2*i]);
+			pfData[i]=(float)tmpdata[i]/L3G_Sensitivity_250dps;
+			// divide by sensitivity
+    }
+		
+		// 28h, 29h -> x axis angular rate
+		// 2Ah, 2Bh -> y axis angular rate
+}
+	
+
+uint8_t MEMS () 
+{
+	int i;
+	for (i=0; i<1000; i++) {
+		
+		read_ang_rate(Buffer);
+		
+		if (Buffer[0]>20 || Buffer[0]<-20) {
+			xangle+=Buffer[0]/100000;
+		}
+		if (Buffer[1]>20 || Buffer[1] <-20) {
+			yangle+=Buffer[1]/100000;
+		}
+	}
+	
+	
+	if (xangle > 10) {
+		return 0x01;
+	}
+	else if (xangle < -10){
+		return 0x02;
+	}
+	else {
+		if (yangle > 10) {
+			return 0x04;
+		}
+		else if (yangle < -10) {
+			return 0x08;
+		}
+		else {
+			return 0x00;
+		}
+	}
+}
```

Comment:
Što točno predstavljaju ove vrijednosti? Ako postoji šansa da će ih u budućnosti trebati mijenjati ili ako imaju neku ulogu u logici možda i njih nebi bilo loše definirati ko konstante?

Reply:
ovaj input.c cu maknuti ovdje je ostala prva verzija funkcije, sve sto je potrebno je u gyro.c.